http-response/background.js

```diff
@@ -1,22 +1,154 @@
 function listener(details) {
+  // The received data is a stream of bytes. In order to do text-based
+  // modifications, it is necessary to decode the bytes into a string
+  // using the proper character encoding, do any modifications, then
+  // encode back into a stream of bytes.
+  // Historically, detecting character encoding has been a tricky task
+  // taken on by the browser. Here, a simplified approach is taken
+  // and the complexity is hidden in a helper method.
+  let decoder, encoder;
+  [decoder, encoder] = detectCharsetAndSetupDecoderEncoder(details);
   let filter = browser.webRequest.filterResponseData(details.requestId);
-  let decoder = new TextDecoder("utf-8");
-  let encoder = new TextEncoder();
+  let fullStr = '';
+  
+  filter.ondata = e => {
+    // Note that the event's data may break in the middle of an encoded
+    // character - the stream parameter is critical for success as this
+    // method gets called multiple times.
+    let str = decoder.decode(e.data, {stream: true});
+    fullStr += str;
+  }
+  
+  filter.onstop = async e => {
+    fullStr += decoder.decode(); //Flush the buffer
+    // Just change any instance of Example or Test in the HTTP response
+    // to WebExtension Example or WebExtension Test.
+    let mutatedStr = fullStr.replace(/Example/g, 'WebExtension Example');
+    mutatedStr = mutatedStr.replace(/Test/g, 'WebExtension Test');
+    filter.write(encoder.encode(mutatedStr));
+    filter.close();
+  }
 
-  filter.ondata = event => {
-    let str = decoder.decode(event.data, {stream: true});
-    // Just change any instance of Example in the HTTP response
-    // to WebExtension Example.
-    str = str.replace(/Example/g, 'WebExtension Example');
-    filter.write(encoder.encode(str));
-    filter.disconnect();
+  filter.onerror = e => {
+    try {
+        filter.close();
+        console.log('Filter error: '+e+', '+ex);
+    } catch(ex) {
+        console.log('Filter error while closing: '+e+', '+ex);
+    }
   }
 
-  return {};
+  // Because details response headers have been mutated, return it
+  return details;
 }
 
-browser.webRequest.onBeforeRequest.addListener(
+browser.webRequest.onHeadersReceived.addListener(
   listener,
-  {urls: ["https://example.com/*"], types: ["main_frame"]},
-  ["blocking"]
+  {
+    urls: ["https://example.com/*", "https://www.w3.org/*"], // Include W3 for testing charset detection.
+    types: ["main_frame"]
+  },
+  ["blocking","responseHeaders"]
 );
+
+// This helper method does a few things regarding character encoding:
+// 1) Detects the charset for the TextDecoder so that bytes are properly turned into strings
+// 2) Ensures the output Content-Type is UTF-8 because that is what TextEncoder supports
+// 3) Returns the decoder/encoder pair
+function detectCharsetAndSetupDecoderEncoder(details) {

```

Comment:
"detect and set up decoder/encoder" suggests that it just does detection and creating new instances of the decoder/encoder.

The method however also mutates the response headers.

Reply:
Yes, I don't like it either. I believe that the real issue here is that TextEncoder does not support anything other than UTF-8, so we **must** force the content type on the output, which is awkward. As noted in the PR description, the code would benefit from having TextEncoder support more types to avoid this hack. Can you think of a way to preserve the original content encoding without the TextEncoder polyfill? I'd rather bypass the mutation altogether because it seems a bit treacherous.

Reply:
For search keywords that are compatible with the used encodings (e.g. letters from the latin alphabet have the same bytes in utf-8, iso-8859-2, ..), you don't need to worry about using TextEncoder. For fixed-length search terms, you could convert the string to bytes and search and replace on typed arrays instead of strings. This may be even more performant than converting to strings and back.

But for the example, I think that it is fine to convert to UTF-8. First, it shows the significance of character encodings.
For those who are not that well versed with character encodings, it is a safe default to choose, because utf-8 is very common.